crates/pathfinder/src/state.rs

```diff
@@ -1,80 +1,438 @@
+        .get(update.address)
+        .context("Get contract state hash from global state tree")?;
+    let contract_root = ContractsStateTable::get_root(db, contract_state_hash)
+        .context("Read contract root from contracts state table")?
+        .unwrap_or(ContractRoot(StarkHash::ZERO));
 
-        Ok(Self { tree })
+    // Load the contract tree and insert the updates.
+    let mut contract_tree =
+        ContractsStateTree::load(db, contract_root).context("Load contract state tree")?;
+    for storage_update in &update.storage_updates {
+        contract_tree
+            .set(storage_update.address, storage_update.value)
+            .context("Update contract storage tree")?;
     }
+    let new_contract_root = contract_tree
+        .apply()
+        .context("Apply contract storage tree changes")?;
 
-    pub fn get(&self, address: ContractAddress) -> anyhow::Result<ContractStateHash> {
-        let value = self.tree.get(address.0)?;
-        Ok(ContractStateHash(value))
-    }
+    // Calculate contract state hash, update global state tree and persist pre-image.
+    let contract_hash = ContractsTable::get_hash(db, update.address)
+        .context("Read contract hash from contracts table")?
+        .context("Contract hash is missing from contracts table")?;
+    let contract_state_hash = calculate_contract_state_hash(contract_hash, new_contract_root);
+
+    ContractsStateTable::insert(db, contract_state_hash, contract_hash, new_contract_root)
+        .context("Insert constract state hash into contracts state table")?;
 
-    pub fn set(
-        &mut self,
-        address: ContractAddress,
-        value: ContractStateHash,
-    ) -> anyhow::Result<()> {
-        self.tree.set(address.0, value.0)
+    Ok(contract_state_hash)
+}
+
+/// Inserts a newly deployed Starknet contract into [ContractsTable].
+async fn deploy_contract(
+    contract: DeployedContract,
+    db: &Transaction<'_>,
+    sequencer: &sequencer::Client,
+) -> anyhow::Result<()> {
+    // Download code and ABI from the sequencer.
+    let code = sequencer
+        .code(
+            H256(contract.address.0.to_be_bytes()),
+            BlockHashOrTag::Tag(Tag::Latest),
+        )
+        .await
+        .context("Download contract code and ABI from sequencer")?;
+
+    // TODO: verify contract hash (waiting on contract definition API change).
+
+    let byte_code = code
+        .bytecode
+        .into_iter()
+        .flat_map(|bytes32| bytes32.0.into_iter())
+        .collect::<Vec<u8>>();
+
+    // TODO: Unsure on how to encode / decode this reliably.
+    let abi = "todo".as_bytes();
+    // TODO: This is not available from sequencer yet.
+    let definition = "does not exist".as_bytes();
+
+    ContractsTable::insert(
+        db,
+        contract.address,
+        contract.hash,
+        &byte_code,
+        abi,
+        definition,
+    )
+    .context("Inserting contract information into contracts table")?;
```

Comment:
Right now the code is mixing async and blocking actions within async context.

It might be easier for the update side to work completly blocking, and only reach out to the runtime for the required downloads?

But the deployed contracts do complicate this. One way would be to:
- on async side, stalk the L1 + download deployed contracts, then send message over mpsc
- on sync side, loop on the channel to *apply* the updates, then send message over mpsc to signal commit (or have a oneshot sender be part of the message)

This would sort of give us some metrics to watch out for (queue size).

An issue is the large definitions to be downloaded, which might use most of the memory as this would need to fully buffer everything.